### PR TITLE
Fix defer'd Close on writer

### PR DESCRIPTION
**Pull request by krehermann**
https://smartcontract-it.atlassian.net/browse/BCF-1354

Add a utility that should be used throughout the repo to wrap defer'd WriteCloser's

The key point is that is safe to call Close on this utility multiple times

The size of this PR grew because there were no test for the nurse, which was the biggest offender of the defer Close problem. Adding tests to satisfy sonarqube uncovered bugs in the implementation opportunities to simplify the code.

### CODE REVIEW

**Review by essamhassan — APPROVED** [review]

File: core/utils/deferable_writer_closer.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+//
+// f, err := os.Create(...)
+// if err != nil {...}
+// wc := NewDeferableWriterCloser(f)
+// defer wc.Close()
+// wc.Write(...)
+// ...
+// err = wc.Close()
+// if err != nil {...}
+// return
```
Comment (jmank88) [inline core/utils/deferable_writer_closer.go:32]: nit/ could indent these code blocks
new go docs has more link support too, if you wish: https://go.dev/doc/comment#links

File: core/utils/deferable_writer_closer.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
+package utils
+
+import (
+	"io"
+	"sync"
+)
+
+// DeferableWriterCloser is to be used in leiu of defer'ing
+// Close on an io.WriterClose (// For more background see https://www.joeshaw.org/dont-defer-close-on-writable-files/)
```
Comment (jmank88) [inline core/utils/deferable_writer_closer.go:9]:
```suggestion
// Close on an io.WriteCloser (// For more background see https://www.joeshaw.org/dont-defer-close-on-writable-files/)
```

File: core/utils/deferable_writer_closer.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+// f, err := os.Create(...)
+// if err != nil {...}
+// wc := NewDeferableWriterCloser(f)
+// defer wc.Close()
+// wc.Write(...)
+// ...
+// err = wc.Close()
+// if err != nil {...}
+// return
+type DeferableWriterCloser struct {
```
Comment (jmank88) [inline core/utils/deferable_writer_closer.go:33]: nit/ maybe consider aligning with `io.WriteCloser` by dropping the 'r'?
```suggestion
type DeferableWriteCloser struct {
```

File: core/utils/deferable_writer_closer.go (hunk 1 above)
Comment (krehermann) [reply core/utils/deferable_writer_closer.go:32]: done

File: core/utils/deferable_write_closer.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
+package utils
+
+import (
+	"io"
+	"sync"
+)
+
+// DeferableWriteCloser is to be used in leiu of defer'ing
+// Close on an io.WriteClose (// For more background see https://www.joeshaw.org/dont-defer-close-on-writable-files/)
```
Comment (jmank88) [inline core/utils/deferable_write_closer.go:9]:
```suggestion
// Close on an [io.WriteCloser] (// For more background see https://www.joeshaw.org/dont-defer-close-on-writable-files/)
```

**Review by jmank88 — APPROVED** [review]

**Pull request merged**